<a href="https://colab.research.google.com/github/zhus-dika/togyz-qumalaq-agent/blob/main/togyzqumalaq_aec_vs_random_policy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  🐘 AEC environment https://pettingzoo.farama.org/api/aec/#about-aec

###  🦜 Install need packages

In [1]:
!pip install gymnasium
!pip install pettingzoo[classic]==1.23.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 953.9/953.9 kB 6.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 826.8/826.8 kB 4.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.1/147.1 kB 4.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 251.1/251.1 kB 6.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 24.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.5/115.5 kB 11.4 MB/s eta 0:00:00
  Created wheel for rlcard: filename=rlcard-1.0.5-py3-none-any.whl size=307097 sha256=765aac62eb17e4c04bdd329818b282aa9d86ce90f5cb3d727a7b7bf4f0781f49
  Stored in directory: /root/.cache/pip/wheels/22/1b/a9/cb9a2009aefc3e311979e3764bc8e095cc6a2541fdc3ce1f60
Successfully built rlcard
  Attempting uninstall: pygame
    Found existing installation: pygame 2.5.2
    Uninstalling pygame-2.5.2:
      Successfully uninstalled pygame-2.5.2


### 🐞 Imports

In [2]:
import functools

import gymnasium
import numpy as np
from gymnasium.spaces import Discrete, MultiDiscrete
from gymnasium import spaces

from pettingzoo import AECEnv
from pettingzoo.utils import agent_selector, wrappers
import matplotlib.pyplot as plt

NUM_ITERS = 400
PLAYS = {"bastaushy": 0, "qostaushy": 0}

### 🦉 Create environment

In [57]:
def env(render_mode=None):
    """
    The env function often wraps the environment in wrappers by default.
    You can find full documentation for these methods
    elsewhere in the developer documentation.
    """
    internal_render_mode = render_mode if render_mode != "ansi" else "human"
    env = raw_env(render_mode=internal_render_mode)
    # This wrapper is only for environments which print results to the terminal
    if render_mode == "ansi":
        env = wrappers.CaptureStdoutWrapper(env)
    # this wrapper helps error handling for discrete action spaces
    env = wrappers.AssertOutOfBoundsWrapper(env)
    # Provides a wide vareity of helpful user errors
    # Strongly recommended
    env = wrappers.OrderEnforcingWrapper(env)
    return env


class raw_env(AECEnv):
    """
    The metadata holds environment constants. From gymnasium, we inherit the "render_modes",
    metadata which specifies which modes can be put into the render() method.
    At least human mode should be supported.
    The "name" metadata allows the environment to be pretty printed.
    """

    metadata = {
        "render_modes": ["ansi", "human"],
        "name": "togyzqumalaq_v0"
        }

    def __init__(self, render_mode=None):
        """
        The init method takes in environment arguments and
         should define the following attributes:
        - otaular
        - tuzdyq
        - qazandar
        - possible_agents
        - render_mode

        Note: as of v1.18.1, the action_spaces and observation_spaces attributes are deprecated.
        Spaces should be defined in the action_space() and observation_space() methods.
        If these methods are not overridden, spaces will be inferred from self.observation_spaces/action_spaces, raising a warning.

        These attributes should not be changed after initialization.
        """
        self.otaular = []
        self.tuzdyq = []
        self.qazandar = []
        self.direction = []
        self.agents = ["bastaushy", "qostaushy"]
        self.possible_agents = self.agents[:]
        # optional: we can define the observation and action spaces here as attributes to be used in their corresponding methods
        self.action_spaces = {i: spaces.Discrete(9) for i in self.agents}
        self.observation_spaces = {
            i: spaces.Dict(
                {
                    "observation": MultiDiscrete([100] * 18 + [9] * 2 + [82] * 2),
                    "action_mask": Discrete(9),
                }
            )
            for i in self.agents
        }
        self.render_mode = render_mode

    # Observation space should be defined here.
    def action_space(self, agent):
        return self.action_spaces[agent]

    # Action space should be defined here.
    def observation_space(self, agent):
        return self.observation_spaces[agent]

    def render(self):
        """
        Renders the environment. In human mode, it can print to terminal, open
        up a graphical window, or open up some other display that a human can see and understand.
        """
        """Renders the environment."""
        if self.render_mode is None:
            gymnasium.logger.warn(
                "You are calling render method without specifying any render mode."
            )
            return

        if len(self.agents) == 2:
            points_bastaushy_x = np.array([i * 2 for i in range(10)])
            points_bastaushy_y = np.array([i % 5 for i in range(50)])

            x = np.arange(-3, 225, 1)
            y = -1

            text_kwargs = dict(ha='center', va='center', fontsize=12)
            plt.figure(figsize=(17, 6))

            for i in range(9):
                # qostaushy's part
                plt.scatter(np.repeat(points_bastaushy_x + 25 * i, 5)[:self.otaular[17 - i]], points_bastaushy_y[:self.otaular[17 - i]], marker='o')
                # horizontal line
                plt.plot(x, np.repeat(y, len(x)))
                # vertical lines
                plt.plot(np.repeat(25 * i - 2, len(x)), np.arange(-7, 5, 12 / len(x)))
                # bastaushy's part
                plt.scatter(np.repeat(points_bastaushy_x + 25 * i, 5)[:self.otaular[i]], points_bastaushy_y[:self.otaular[i]] - 6, marker='o')

            #last vertical line
            plt.plot(np.repeat(25 * 9 - 2, len(x)), np.arange(-7, 5, 12 / len(x)))

            for i in range(9):
                # bastaushy's qumalaqtar
                plt.text(25 * i + 10, -7, f'{i} ({self.otaular[i]})', **text_kwargs)
                # qostaushy's qumalaqtar
                plt.text(25 * i + 10, 5, f'{17 - i} ({self.otaular[17 - i]})', **text_kwargs)
            # bastaushy qazan's qumalaqtar
            plt.text(230, -4, f'qazan: {self.qazandar[0]}', **text_kwargs)
            # qostaushy qazan's qumalaqtar
            plt.text(230, 2, f'qazan: {self.qazandar[1]}', **text_kwargs);
            # bastaushy tuzdyq's qumalaqtar
            plt.text(230, -6, f'tuzdyq: {self.tuzdyq[0]}', **text_kwargs)
            # qostaushy tuzdyq's qumalaqtar
            plt.text(230, 0, f'tuzdyq: {self.tuzdyq[1]}', **text_kwargs);
            plt.show()
        else:
            if self.render_mode == "human":
                print("Game over")

    def _legal_moves(self, agent):
        cur_player = self.possible_agents.index(agent)
        opp_player = (cur_player + 1) % 2
        return [item for item in range(9 * cur_player, (cur_player + 1) * 9) if self.tuzdyq[opp_player] != item and self.otaular[item] > 0]

    def observe(self, agent):
        """
        Observe should return the observation of the specified agent. This function
        should return a sane observation (though not necessarily the most up to date possible)
        at any time after reset() is called.
        """
        # observation of one agent is the previous state of the other
        legal_moves = self._legal_moves(agent) if agent == self.agent_selection else []
        action_mask = np.zeros(9, "int8")
        if self.possible_agents.index(agent) == 1:
            legal_moves = [i - 9 for i in legal_moves]
        for i in legal_moves:
            action_mask[i] = 1
        observation = tuple(
            self.otaular + self.tuzdyq + self.qazandar
        )
        return {"observation": observation, "action_mask": action_mask}

    def close(self):
        """
        Close should release any graphical displays, subprocesses, network connections
        or any other environment data which should not be kept around after the
        user is no longer using the environment.
        """
        pass

    def reset(self, seed=None, options=None):
        """
        Reset needs to initialize the following attributes
        - agents
        - rewards
        - _cumulative_rewards
        - terminations
        - truncations
        - infos
        - agent_selection
        And must set up the environment so that render(), step(), and observe()
        can be called without issues.
        Here it sets up the state dictionary which is used by step() and the observations dictionary which is used by step() and observe()
        """
        self.agents = self.possible_agents[:]
        self.rewards = {agent: 0 for agent in self.agents}
        self._cumulative_rewards = {agent: 0 for agent in self.agents}
        self.otaular = [9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9]
        self.direction = [list(range(18)), [9, 10, 11, 12, 13, 14, 15, 16, 17, 0, 1, 2, 3, 4, 5, 6, 7, 8]]
        self.tuzdyq = [-1, -1]
        self.qazandar = [0, 0]
        self.terminations = {agent: False for agent in self.agents}
        self.truncations = {agent: False for agent in self.agents}
        self.infos = {agent: {} for agent in self.agents}
        self.num_moves = 0
        observation = tuple(
            self.otaular + self.tuzdyq + self.qazandar
        )
        self.observations = {agent: observation for agent in self.agents}
        """
        Our agent_selector utility allows easy cyclic stepping through the agents list.
        """
        self._agent_selector = agent_selector(self.agents)
        self.agent_selection = self._agent_selector.next()

    def step(self, action):
        """
        step(action) takes in an action for the current agent (specified by
        agent_selection) and needs to update
        - rewards
        - _cumulative_rewards (accumulating the rewards)
        - terminations
        - truncations
        - infos
        - agent_selection (to the next agent)
        And any internal state used by observe() or render()
        """
        if (
            self.terminations[self.agent_selection]
            or self.truncations[self.agent_selection]
        ):
            # handles stepping an agent which is already dead
            # accepts a None action for the one agent, and moves the agent_selection to
            # the next dead agent,  or if there are no more dead agents, to the next live agent
            self._was_dead_step(action)
            return

        cur_player = self.possible_agents.index(self.agent_selection)
        opp_player = (cur_player + 1) % 2
        self.num_moves += 1
        if self.render_mode == "human":
            print(f'MOVE #{self.num_moves}')
        # The truncations dictionary must be updated for all players.
        self.truncations = {
            agent: self.num_moves >= NUM_ITERS for agent in self.agents
        }
        # distribute qumalaqs
        if cur_player == 1:
            action += 9
        if self.render_mode == "human":
            print(f'{self.agent_selection} made action {action}')
        num_qumalaq = self.otaular[action]
        idx_action = self.direction[cur_player].index(action)
        if self.otaular[action] == 1:
            self.otaular[self.direction[cur_player][idx_action + 1]] += 1
            self.otaular[action] -= 1
        else:
            i = 1
            while self.otaular[action] > 1:
                self.otaular[self.direction[cur_player][(idx_action + i) % 18]] += 1
                self.otaular[action] -= 1
                i += 1
        # check tuzdyq & add rewards to qazandar
        reward = 0
        if self.check_tuzdyq(self.agent_selection, action):
            reward += 3
            if self.render_mode == "human":
                print(f'{self.agent_selection} won tuzdyq {reward}')
        else:

            if num_qumalaq > 1:
                last_otau = self.direction[cur_player][(idx_action + num_qumalaq - 1) % 18]
            else:
                last_otau = self.direction[cur_player][(idx_action + num_qumalaq) % 18]

            if last_otau in range(opp_player * 9, (opp_player + 1) * 9) and self.otaular[last_otau] % 2 == 0:
                reward += self.otaular[last_otau]
                if self.render_mode == "human":
                    print(f'{self.agent_selection} won {reward}')
                self.otaular[last_otau] = 0
            if self.tuzdyq[cur_player] >= 0 and self.otaular[self.tuzdyq[cur_player]] > 0:
                reward += self.otaular[self.tuzdyq[cur_player]]
                if self.render_mode == "human":
                    print(f'{self.agent_selection} won tuzdyq {self.otaular[self.tuzdyq[cur_player]]}')
                self.otaular[self.tuzdyq[cur_player]] = 0
        if self.render_mode == "human":
            print(f'{self.agent_selection} won total {reward}')
        self.qazandar[cur_player] += reward
        self.rewards[self.agent_selection] += reward
        # check if there is a winner
        winner = self.check_for_winner()
        if winner:
            self.terminations = {i: True for i in self.agents}
            if self.render_mode == "human":
                print(f'{self.agent_selection} won the game!!!')
        # selects the next agent.
        self.agent_selection = self._agent_selector.next()
        # Adds .rewards to ._cumulative_rewards
        self._accumulate_rewards()

        if self.render_mode == "human":
            self.render()

    def check_tuzdyq(self, agent, action):
        cur_player = self.possible_agents.index(agent)
        opp_player = (cur_player + 1) % 2
        idx = self.direction[cur_player].index(action)
        num_qumalaq = self.otaular[action]

        if num_qumalaq > 1:
            last_otau = self.direction[cur_player][(idx + num_qumalaq - 1) % 18]
        else:
            last_otau = self.direction[cur_player][(idx + num_qumalaq) % 18]

        if last_otau in range(opp_player * 9, (opp_player + 1) * 9) and self.otaular[last_otau] == 3 and last_otau != 17 - cur_player * 9 and abs(last_otau - self.tuzdyq[opp_player]) != 9:
            self.tuzdyq[cur_player] = last_otau
            self.otaular[last_otau] = 0
            if self.render_mode == "human":
                print(f'{agent} got tuzdyq {last_otau}!')
            return True

        return False

    def check_atsyrau(self, agent):
        cur_player = self.possible_agents.index(agent)
        opp_player = (cur_player + 1) % 2

        for idx, i in enumerate(self.otaular[cur_player * 9: (cur_player + 1) * 9]):
            if i > 0 and idx + cur_player * 9 != self.tuzdyq[opp_player]:
                return False
        if self.render_mode == "human":
            print(f'{agent} reached atsyrau')
        return True

    def check_for_winner(self):
        cur_player = self.possible_agents.index(self.agent_selection)
        opp_player = (cur_player + 1) % 2
        if self.qazandar[cur_player] > 81:
            PLAYS[self.agent_selection] += 1
            return True
        if self.check_atsyrau(self.possible_agents[opp_player]) and self.qazandar[opp_player] <= 81:
            PLAYS[self.agent_selection] += 1
            return True
        return False

### 🦚 Testing environment

In [ ]:
# env = env(render_mode="human")
# env.reset(seed=42)

# for agent in env.agent_iter():
#     observation, reward, termination, truncation, info = env.last()

#     if termination or truncation:
#         action = None
#     else:
#         mask = observation["action_mask"]
#         # this is where you would insert your policy
#         action = env.action_space(agent).sample(mask)

#     env.step(action)
# env.close()

# 🐼 DQN agent to play vs a random policy agent https://pettingzoo.farama.org/tutorials/tianshou/intermediate/

### 🐡 Additional installations https://github.com/thu-ml/tianshou

In [4]:
!pip install packaging==21.3
!pip install tianshou==0.5.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.8/40.8 kB 1.5 MB/s eta 0:00:00
  Attempting uninstall: packaging
    Found existing installation: packaging 24.0
    Uninstalling packaging-24.0:
      Successfully uninstalled packaging-24.0


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 162.7/162.7 kB 2.5 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.19.3-py3-none-manylinux1_x86_64.whl (166.0 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-many

### 🐝 Imports

In [5]:
import os
from typing import Optional, Tuple

import gymnasium
import numpy as np
import torch
from copy import deepcopy
from tianshou.data import Collector, VectorReplayBuffer
from tianshou.env import DummyVectorEnv
from tianshou.env.pettingzoo_env import PettingZooEnv
from tianshou.policy import BasePolicy, DQNPolicy, RainbowPolicy, MultiAgentPolicyManager, RandomPolicy
from tianshou.trainer import offpolicy_trainer
from tianshou.utils.net.common import Net

### 🐫 Prepare main functions

In [6]:
agent1_path = "policy_128x256x256x128_bs64.pth"
agent2_path = "policy_256x512x512x256_bs128.pth"
agent3_path = "policy_512x1024x1024x512_bs128.pth"

In [ ]:
def _get_agents(
    agent_learn: Optional[BasePolicy] = None,
    agent_opponent: Optional[BasePolicy] = None,
    optim: Optional[torch.optim.Optimizer] = None,
) -> Tuple[BasePolicy, torch.optim.Optimizer, list]:
    env = _get_env()
    observation_space = (
        env.observation_space["observation"]
        if isinstance(env.observation_space, gymnasium.spaces.Dict)
        else env.observation_space
    )
    if agent_learn is None:
        # model
        net = Net(
            state_shape=observation_space.shape or observation_space.n,
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[512, 1024, 1024, 512],
            #hidden_sizes=[1024, 2048, 2048, 1024],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

        if optim is None:
            optim = torch.optim.Adam(net.parameters(), lr=1e-4)
        agent_learn = DQNPolicy(
            model=net,
            optim=optim,
            discount_factor=0.9,
            estimation_step=3,
            target_update_freq=320,
        ).to("cuda" if torch.cuda.is_available() else "cpu")


    if agent_opponent is None:
        if opponent_path:
            agent_opponent = deepcopy(agent_learn)
            agent_opponent.load_state_dict(torch.load(opponent_path))
        else:
            agent_opponent = RandomPolicy(action_space=env.action_space)

    agents = [agent_opponent, agent_learn]
    #agents = [agent_learn, agent_opponent]
    policy = MultiAgentPolicyManager(agents, env)
    return policy, optim, env.agents


def _get_env():
    """This function is needed to provide callables for DummyVectorEnv."""
    return PettingZooEnv(env())

###  🐑 Training code https://tianshou.org/en/stable/01_tutorials/04_tictactoe.html

In [ ]:
# Before evaluate this cell run the cell with env

# ======== Step 1: Environment setup =========
train_envs = DummyVectorEnv([_get_env for _ in range(100)])
test_envs = DummyVectorEnv([_get_env for _ in range(100)])

# seed
seed = 777
np.random.seed(seed)
torch.manual_seed(seed)
train_envs.seed(seed)
test_envs.seed(seed)

# ======== Step 2: Agent setup =========
policy, optim, agents = _get_agents()

# # ======== Step 3: Collector setup =========
train_collector = Collector(
    policy,
    train_envs,
    VectorReplayBuffer(20_000, len(train_envs)),
    exploration_noise=True,
)
test_collector = Collector(policy, test_envs, exploration_noise=True)
# policy.set_eps(1)
train_collector.collect(n_step=128 * 100)  # batch size * training_num

# ======== Step 4: Callback functions setup =========
def save_best_fn(policy):
    model_save_path = os.path.join("log", "ttt", "dqn", "policy.pth")
    os.makedirs(os.path.join("log", "ttt", "dqn"), exist_ok=True)
    torch.save(policy.policies[agents[1]].state_dict(), model_save_path)

def stop_fn(mean_rewards):
    return mean_rewards >= 21000

def train_fn(epoch, env_step):
    policy.policies[agents[1]].set_eps(0.1)

def test_fn(epoch, env_step):
    policy.policies[agents[1]].set_eps(0.05)

def reward_metric(rews):
    return rews[:, 1]

# ======== Step 5: Run the trainer =========
result = offpolicy_trainer(
    policy=policy,
    train_collector=train_collector,
    test_collector=test_collector,
    max_epoch=100,
    step_per_epoch=1000,
    step_per_collect=50,
    episode_per_test=10,
    batch_size=128,
    train_fn=train_fn,
    test_fn=test_fn,
    stop_fn=stop_fn,
    save_best_fn=save_best_fn,
    update_per_step=0.1,
    test_in_train=False,
    reward_metric=reward_metric,
    verbose=True
)

# return result, policy.policies[agents[1]]
print(f"\n==========Result==========\n{result}")
print("\n(the trained policy can be accessed via policy.policies[agents[1]])")

Epoch #1: 1001it [00:12, 80.67it/s, bastaushy/loss=270.057, env_step=1000, len=0, n/ep=0, n/st=100, qostaushy/loss=2765.895, rew=0.00]                          


Epoch #1: test_reward: 10799.400000 ± 3947.835564, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #2: 1001it [00:10, 97.18it/s, bastaushy/loss=111.903, env_step=2000, len=148, n/ep=1, n/st=100, qostaushy/loss=136.716, rew=5858.00]                           


Epoch #2: test_reward: 12951.400000 ± 2550.261524, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #3: 1001it [00:10, 99.00it/s, bastaushy/loss=201.262, env_step=3000, len=154, n/ep=0, n/st=100, qostaushy/loss=93.804, rew=6295.00]                           


Epoch #3: test_reward: 12856.100000 ± 5922.258428, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #4: 1001it [00:09, 101.29it/s, bastaushy/loss=473.575, env_step=4000, len=166, n/ep=0, n/st=100, qostaushy/loss=524.497, rew=7274.00]                          


Epoch #4: test_reward: 13995.600000 ± 4737.990950, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #5: 1001it [00:09, 100.59it/s, bastaushy/loss=598.521, env_step=5000, len=170, n/ep=0, n/st=100, qostaushy/loss=222.464, rew=6634.60]                          


Epoch #5: test_reward: 14260.900000 ± 2920.463369, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #6: 1001it [00:10, 98.88it/s, bastaushy/loss=816.723, env_step=6000, len=180, n/ep=0, n/st=100, qostaushy/loss=275.041, rew=7041.00]                           


Epoch #6: test_reward: 11720.700000 ± 4819.362324, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #7: 1001it [00:10, 93.40it/s, bastaushy/loss=883.190, env_step=7000, len=180, n/ep=0, n/st=100, qostaushy/loss=585.974, rew=7041.00]                          


Epoch #7: test_reward: 9399.500000 ± 2530.931261, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #8: 1001it [00:11, 84.99it/s, bastaushy/loss=684.352, env_step=8000, len=180, n/ep=0, n/st=100, qostaushy/loss=536.661, rew=7041.00]                          


Epoch #8: test_reward: 8197.900000 ± 4291.245611, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #9: 1001it [00:12, 79.36it/s, bastaushy/loss=735.293, env_step=9000, len=180, n/ep=0, n/st=100, qostaushy/loss=653.424, rew=7041.00]                          


Epoch #9: test_reward: 15532.700000 ± 4497.339237, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #10: 1001it [00:11, 83.66it/s, bastaushy/loss=763.006, env_step=10000, len=180, n/ep=0, n/st=100, qostaushy/loss=936.676, rew=7041.00]                          


Epoch #10: test_reward: 10097.200000 ± 3728.826137, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #11: 1001it [00:12, 83.29it/s, bastaushy/loss=455.495, env_step=11000, len=180, n/ep=0, n/st=100, qostaushy/loss=978.914, rew=7041.00]                          


Epoch #11: test_reward: 12497.200000 ± 4128.103991, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #12: 1001it [00:12, 77.26it/s, bastaushy/loss=626.474, env_step=12000, len=244, n/ep=0, n/st=100, qostaushy/loss=767.202, rew=11362.00]                          


Epoch #12: test_reward: 9101.500000 ± 3260.406179, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #13: 1001it [00:12, 80.87it/s, bastaushy/loss=639.062, env_step=13000, len=257, n/ep=0, n/st=100, qostaushy/loss=1180.969, rew=11872.00]                          


Epoch #13: test_reward: 10155.800000 ± 3471.003826, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #14: 1001it [00:12, 80.01it/s, bastaushy/loss=981.792, env_step=14000, len=268, n/ep=1, n/st=100, qostaushy/loss=1955.858, rew=12994.00]                          


Epoch #14: test_reward: 10373.600000 ± 4874.737207, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #15: 1001it [00:12, 78.94it/s, bastaushy/loss=989.711, env_step=15000, len=278, n/ep=2, n/st=100, qostaushy/loss=2034.027, rew=13697.00]                          


Epoch #15: test_reward: 9822.800000 ± 4432.049928, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #16: 1001it [00:12, 77.69it/s, bastaushy/loss=1408.390, env_step=16000, len=108, n/ep=1, n/st=100, qostaushy/loss=1664.387, rew=4890.00]                          


Epoch #16: test_reward: 3575.400000 ± 2611.969992, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #17: 1001it [00:12, 79.23it/s, bastaushy/loss=1312.736, env_step=17000, len=150, n/ep=1, n/st=100, qostaushy/loss=2415.719, rew=9360.00]                          


Epoch #17: test_reward: 14613.800000 ± 6818.803734, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #18: 1001it [00:12, 79.23it/s, bastaushy/loss=1431.187, env_step=18000, len=120, n/ep=0, n/st=100, qostaushy/loss=2831.133, rew=6282.00]                          


Epoch #18: test_reward: 13140.400000 ± 2904.572092, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #19: 1001it [00:12, 78.67it/s, bastaushy/loss=1409.744, env_step=19000, len=229, n/ep=0, n/st=100, qostaushy/loss=2360.154, rew=12944.00]                          


Epoch #19: test_reward: 7379.500000 ± 4258.406844, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #20: 1001it [00:12, 78.66it/s, bastaushy/loss=1657.724, env_step=20000, len=328, n/ep=1, n/st=100, qostaushy/loss=3422.645, rew=17544.00]                          


Epoch #20: test_reward: 12545.400000 ± 6740.766903, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #21: 1001it [00:13, 74.23it/s, bastaushy/loss=1801.802, env_step=21000, len=337, n/ep=0, n/st=100, qostaushy/loss=3687.742, rew=16784.00]                          


Epoch #21: test_reward: 10545.500000 ± 3541.176930, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #22: 1001it [00:12, 78.12it/s, bastaushy/loss=2433.778, env_step=22000, len=348, n/ep=1, n/st=100, qostaushy/loss=3883.549, rew=18844.00]                          


Epoch #22: test_reward: 9799.400000 ± 2893.567494, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #23: 1001it [00:13, 76.90it/s, bastaushy/loss=2677.015, env_step=23000, len=358, n/ep=1, n/st=100, qostaushy/loss=4529.737, rew=19624.00]                          


Epoch #23: test_reward: 10819.200000 ± 4794.280004, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #24: 1001it [00:12, 77.67it/s, bastaushy/loss=2335.423, env_step=24000, len=277, n/ep=0, n/st=100, qostaushy/loss=4580.123, rew=15558.00]                          


Epoch #24: test_reward: 10581.800000 ± 4239.096809, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #25: 1001it [00:12, 77.46it/s, bastaushy/loss=2485.229, env_step=25000, len=197, n/ep=0, n/st=100, qostaushy/loss=5355.015, rew=10176.00]                          


Epoch #25: test_reward: 11015.400000 ± 4182.801267, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #26: 1001it [00:12, 77.39it/s, bastaushy/loss=2781.117, env_step=26000, len=387, n/ep=0, n/st=100, qostaushy/loss=4772.597, rew=22152.00]                          


Epoch #26: test_reward: 10606.200000 ± 3169.226335, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #27: 1001it [00:12, 77.21it/s, bastaushy/loss=2691.468, env_step=27000, len=397, n/ep=0, n/st=100, qostaushy/loss=4088.952, rew=22572.00]                          


Epoch #27: test_reward: 11041.300000 ± 4970.573771, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #28: 1001it [00:12, 77.64it/s, bastaushy/loss=2705.067, env_step=28000, len=147, n/ep=0, n/st=100, qostaushy/loss=5230.582, rew=8956.00]                          


Epoch #28: test_reward: 13388.200000 ± 3260.883156, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #29: 1001it [00:13, 76.32it/s, bastaushy/loss=2741.746, env_step=29000, len=129, n/ep=0, n/st=100, qostaushy/loss=4978.288, rew=5016.00]                          


Epoch #29: test_reward: 10221.600000 ± 4511.669296, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #30: 1001it [00:13, 76.76it/s, bastaushy/loss=2831.437, env_step=30000, len=159, n/ep=1, n/st=100, qostaushy/loss=4258.850, rew=7200.00]                          


Epoch #30: test_reward: 13038.400000 ± 3691.890984, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #31: 1001it [00:13, 72.87it/s, bastaushy/loss=2694.364, env_step=31000, len=180, n/ep=0, n/st=100, qostaushy/loss=5100.143, rew=11610.00]                          


Epoch #31: test_reward: 16332.500000 ± 4730.730879, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #32: 1001it [00:13, 76.04it/s, bastaushy/loss=2730.338, env_step=32000, len=268, n/ep=1, n/st=100, qostaushy/loss=5302.998, rew=15984.00]                          


Epoch #32: test_reward: 13007.800000 ± 5306.173325, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #33: 1001it [00:13, 75.43it/s, bastaushy/loss=2982.550, env_step=33000, len=107, n/ep=0, n/st=100, qostaushy/loss=4736.273, rew=6389.00]                          


Epoch #33: test_reward: 10599.900000 ± 4619.962261, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #34: 1001it [00:13, 76.74it/s, bastaushy/loss=3236.827, env_step=34000, len=216, n/ep=0, n/st=100, qostaushy/loss=4348.962, rew=13831.00]                          


Epoch #34: test_reward: 10623.000000 ± 4280.000327, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #35: 1001it [00:13, 76.73it/s, bastaushy/loss=2804.256, env_step=35000, len=307, n/ep=0, n/st=100, qostaushy/loss=5424.095, rew=17352.00]                          


Epoch #35: test_reward: 12310.900000 ± 3191.417097, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #36: 1001it [00:13, 76.40it/s, bastaushy/loss=3199.747, env_step=36000, len=160, n/ep=1, n/st=100, qostaushy/loss=4579.332, rew=9366.00]                          


Epoch #36: test_reward: 7280.900000 ± 2736.021983, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #37: 1001it [00:13, 76.52it/s, bastaushy/loss=2862.584, env_step=37000, len=237, n/ep=2, n/st=100, qostaushy/loss=4713.579, rew=15526.00]                          


Epoch #37: test_reward: 15030.800000 ± 3705.796130, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #38: 1001it [00:13, 76.39it/s, bastaushy/loss=2898.049, env_step=38000, len=223, n/ep=1, n/st=100, qostaushy/loss=4181.373, rew=12612.00]                          


Epoch #38: test_reward: 8566.100000 ± 3622.231093, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #39: 1001it [00:13, 74.83it/s, bastaushy/loss=2612.066, env_step=39000, len=132, n/ep=1, n/st=100, qostaushy/loss=4791.786, rew=7474.00]                          


Epoch #39: test_reward: 11453.600000 ± 4060.008355, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #40: 1001it [00:13, 72.77it/s, bastaushy/loss=3368.396, env_step=40000, len=220, n/ep=2, n/st=100, qostaushy/loss=4234.125, rew=10512.00]                          


Epoch #40: test_reward: 12798.700000 ± 5812.412048, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #41: 1001it [00:13, 76.76it/s, bastaushy/loss=3264.767, env_step=41000, len=203, n/ep=0, n/st=100, qostaushy/loss=4032.743, rew=12336.00]                          


Epoch #41: test_reward: 13885.700000 ± 2871.938789, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #42: 1001it [00:13, 76.44it/s, bastaushy/loss=2670.233, env_step=42000, len=210, n/ep=0, n/st=100, qostaushy/loss=3927.700, rew=10710.00]                          


Epoch #42: test_reward: 12555.400000 ± 4405.251325, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #43: 1001it [00:13, 76.69it/s, bastaushy/loss=2716.544, env_step=43000, len=147, n/ep=0, n/st=100, qostaushy/loss=3284.738, rew=6155.00]                          


Epoch #43: test_reward: 16349.800000 ± 5013.833779, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #44: 1001it [00:13, 75.43it/s, bastaushy/loss=3080.175, env_step=44000, len=235, n/ep=0, n/st=100, qostaushy/loss=2940.202, rew=14148.00]                          


Epoch #44: test_reward: 10581.200000 ± 1418.388720, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #45: 1001it [00:13, 76.59it/s, bastaushy/loss=3130.203, env_step=45000, len=195, n/ep=1, n/st=100, qostaushy/loss=3403.299, rew=7378.00]                          


Epoch #45: test_reward: 11046.500000 ± 4401.005732, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #46: 1001it [00:13, 76.65it/s, bastaushy/loss=3104.728, env_step=46000, len=257, n/ep=1, n/st=100, qostaushy/loss=3427.163, rew=10972.00]                          


Epoch #46: test_reward: 12548.800000 ± 4233.680923, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #47: 1001it [00:13, 76.44it/s, bastaushy/loss=3109.147, env_step=47000, len=234, n/ep=0, n/st=100, qostaushy/loss=2926.756, rew=9484.00]                          


Epoch #47: test_reward: 15653.100000 ± 5465.432214, best_reward: 16625.700000 ± 4250.630260 in #0


Epoch #48: 1001it [00:13, 76.11it/s, bastaushy/loss=2881.930, env_step=48000, len=98, n/ep=0, n/st=100, qostaushy/loss=3215.308, rew=5240.00]                          


Epoch #48: test_reward: 16744.600000 ± 4604.044683, best_reward: 16744.600000 ± 4604.044683 in #48


Epoch #49: 1001it [00:14, 69.02it/s, bastaushy/loss=3260.083, env_step=49000, len=224, n/ep=1, n/st=100, qostaushy/loss=3693.429, rew=14386.00]                          


Epoch #49: test_reward: 14534.600000 ± 6140.554685, best_reward: 16744.600000 ± 4604.044683 in #48


Epoch #50: 1001it [00:13, 76.08it/s, bastaushy/loss=3244.766, env_step=50000, len=288, n/ep=0, n/st=100, qostaushy/loss=3128.700, rew=18776.00]                          


Epoch #50: test_reward: 17288.600000 ± 5902.201813, best_reward: 17288.600000 ± 5902.201813 in #50


Epoch #51: 1001it [00:13, 76.01it/s, bastaushy/loss=3081.336, env_step=51000, len=212, n/ep=0, n/st=100, qostaushy/loss=3460.241, rew=8507.00]                          


Epoch #51: test_reward: 14026.700000 ± 6472.901761, best_reward: 17288.600000 ± 5902.201813 in #50


Epoch #52: 1001it [00:13, 76.06it/s, bastaushy/loss=3416.710, env_step=52000, len=161, n/ep=0, n/st=100, qostaushy/loss=3494.780, rew=7452.00]                          


Epoch #52: test_reward: 20748.800000 ± 2072.252195, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #53: 1001it [00:13, 75.87it/s, bastaushy/loss=2654.459, env_step=53000, len=190, n/ep=0, n/st=100, qostaushy/loss=2582.663, rew=10834.00]                          


Epoch #53: test_reward: 15457.900000 ± 5498.025290, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #54: 1001it [00:13, 74.73it/s, bastaushy/loss=3385.670, env_step=54000, len=280, n/ep=0, n/st=100, qostaushy/loss=2538.109, rew=16726.50]                          


Epoch #54: test_reward: 13023.300000 ± 5589.134997, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #55: 1001it [00:13, 75.85it/s, bastaushy/loss=3291.726, env_step=55000, len=287, n/ep=0, n/st=100, qostaushy/loss=2410.184, rew=16756.00]                          


Epoch #55: test_reward: 12000.200000 ± 7099.538208, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #56: 1001it [00:13, 75.56it/s, bastaushy/loss=2927.545, env_step=56000, len=310, n/ep=0, n/st=100, qostaushy/loss=2941.991, rew=21228.00]                          


Epoch #56: test_reward: 15300.300000 ± 6729.148075, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #57: 1001it [00:13, 75.84it/s, bastaushy/loss=2678.108, env_step=57000, len=244, n/ep=1, n/st=100, qostaushy/loss=2473.068, rew=16056.00]                          


Epoch #57: test_reward: 15055.100000 ± 6612.225532, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #58: 1001it [00:14, 70.57it/s, bastaushy/loss=3294.424, env_step=58000, len=160, n/ep=0, n/st=100, qostaushy/loss=3074.328, rew=10499.00]                          


Epoch #58: test_reward: 13352.400000 ± 6877.931334, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #59: 1001it [00:13, 75.86it/s, bastaushy/loss=2299.836, env_step=59000, len=173, n/ep=0, n/st=100, qostaushy/loss=3001.926, rew=9232.00]                          


Epoch #59: test_reward: 15514.700000 ± 4336.569429, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #60: 1001it [00:13, 75.56it/s, bastaushy/loss=2742.782, env_step=60000, len=217, n/ep=0, n/st=100, qostaushy/loss=2992.686, rew=9420.00]                          


Epoch #60: test_reward: 8210.100000 ± 3280.578438, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #61: 1001it [00:13, 75.58it/s, bastaushy/loss=2619.596, env_step=61000, len=284, n/ep=0, n/st=100, qostaushy/loss=3016.771, rew=17933.00]                          


Epoch #61: test_reward: 9066.700000 ± 2982.469247, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #62: 1001it [00:13, 75.57it/s, bastaushy/loss=2206.060, env_step=62000, len=400, n/ep=1, n/st=100, qostaushy/loss=2397.644, rew=24226.00]                          


Epoch #62: test_reward: 13329.800000 ± 4608.334662, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #63: 1001it [00:13, 74.54it/s, bastaushy/loss=2668.290, env_step=63000, len=181, n/ep=0, n/st=100, qostaushy/loss=2534.332, rew=8492.00]                          


Epoch #63: test_reward: 18716.500000 ± 6916.441300, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #64: 1001it [00:13, 75.34it/s, bastaushy/loss=2326.234, env_step=64000, len=156, n/ep=0, n/st=100, qostaushy/loss=2700.535, rew=8857.00]                          


Epoch #64: test_reward: 15823.300000 ± 4716.217044, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #65: 1001it [00:13, 75.53it/s, bastaushy/loss=2526.801, env_step=65000, len=240, n/ep=0, n/st=100, qostaushy/loss=3368.495, rew=14795.00]                          


Epoch #65: test_reward: 17916.300000 ± 4432.371488, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #66: 1001it [00:13, 75.16it/s, bastaushy/loss=2131.056, env_step=66000, len=244, n/ep=0, n/st=100, qostaushy/loss=2976.709, rew=16516.00]                          


Epoch #66: test_reward: 11499.300000 ± 4424.930328, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #67: 1001it [00:13, 71.51it/s, bastaushy/loss=1837.717, env_step=67000, len=400, n/ep=0, n/st=100, qostaushy/loss=3022.221, rew=23300.00]                          


Epoch #67: test_reward: 9558.400000 ± 4701.440209, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #68: 1001it [00:13, 72.52it/s, bastaushy/loss=2277.522, env_step=68000, len=314, n/ep=1, n/st=100, qostaushy/loss=2997.935, rew=19893.00]                          


Epoch #68: test_reward: 13447.000000 ± 5990.952362, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #69: 1001it [00:13, 74.82it/s, bastaushy/loss=2235.950, env_step=69000, len=234, n/ep=0, n/st=100, qostaushy/loss=3193.916, rew=13405.00]                          


Epoch #69: test_reward: 9714.600000 ± 3464.039036, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #70: 1001it [00:13, 74.44it/s, bastaushy/loss=1547.948, env_step=70000, len=297, n/ep=0, n/st=100, qostaushy/loss=3117.144, rew=18500.00]                          


Epoch #70: test_reward: 14458.900000 ± 4044.083319, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #71: 1001it [00:13, 74.44it/s, bastaushy/loss=2067.592, env_step=71000, len=116, n/ep=1, n/st=100, qostaushy/loss=2928.147, rew=5634.00]                          


Epoch #71: test_reward: 10292.900000 ± 2987.292368, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #72: 1001it [00:13, 74.68it/s, bastaushy/loss=2181.316, env_step=72000, len=228, n/ep=0, n/st=100, qostaushy/loss=2995.118, rew=14959.00]                          


Epoch #72: test_reward: 12555.900000 ± 3485.855259, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #73: 1001it [00:13, 72.52it/s, bastaushy/loss=2009.769, env_step=73000, len=106, n/ep=0, n/st=100, qostaushy/loss=2860.566, rew=6128.00]                          


Epoch #73: test_reward: 10363.700000 ± 3886.847721, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #74: 1001it [00:13, 74.79it/s, bastaushy/loss=1848.589, env_step=74000, len=236, n/ep=0, n/st=100, qostaushy/loss=3166.375, rew=16244.00]                          


Epoch #74: test_reward: 13097.900000 ± 4184.898026, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #75: 1001it [00:13, 74.97it/s, bastaushy/loss=1864.220, env_step=75000, len=257, n/ep=0, n/st=100, qostaushy/loss=3361.660, rew=11212.00]                          


Epoch #75: test_reward: 12967.400000 ± 3978.599256, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #76: 1001it [00:13, 73.92it/s, bastaushy/loss=2048.368, env_step=76000, len=234, n/ep=1, n/st=100, qostaushy/loss=3728.016, rew=14795.00]                          


Epoch #76: test_reward: 12237.600000 ± 4081.394521, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #77: 1001it [00:13, 75.18it/s, bastaushy/loss=2205.551, env_step=77000, len=176, n/ep=2, n/st=100, qostaushy/loss=3835.769, rew=11112.00]                          


Epoch #77: test_reward: 9331.500000 ± 4409.580598, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #78: 1001it [00:13, 75.03it/s, bastaushy/loss=2559.547, env_step=78000, len=177, n/ep=1, n/st=100, qostaushy/loss=2743.137, rew=9314.00]                          


Epoch #78: test_reward: 10208.900000 ± 4280.859130, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #79: 1001it [00:13, 74.85it/s, bastaushy/loss=2329.673, env_step=79000, len=289, n/ep=2, n/st=100, qostaushy/loss=3216.085, rew=18482.00]                          


Epoch #79: test_reward: 12727.300000 ± 5827.241424, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #80: 1001it [00:13, 74.94it/s, bastaushy/loss=2824.043, env_step=80000, len=161, n/ep=0, n/st=100, qostaushy/loss=3655.198, rew=10156.00]                          


Epoch #80: test_reward: 12727.900000 ± 4871.339045, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #81: 1001it [00:13, 75.36it/s, bastaushy/loss=2365.626, env_step=81000, len=177, n/ep=2, n/st=100, qostaushy/loss=3326.921, rew=11059.50]                          


Epoch #81: test_reward: 11605.400000 ± 3269.345353, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #82: 1001it [00:13, 74.42it/s, bastaushy/loss=2473.191, env_step=82000, len=79, n/ep=1, n/st=100, qostaushy/loss=3219.113, rew=3384.00]                          


Epoch #82: test_reward: 10904.000000 ± 4138.195090, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #83: 1001it [00:13, 75.16it/s, bastaushy/loss=2525.048, env_step=83000, len=162, n/ep=2, n/st=100, qostaushy/loss=4004.970, rew=9848.50]                          


Epoch #83: test_reward: 12310.800000 ± 4405.590194, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #84: 1001it [00:13, 75.49it/s, bastaushy/loss=2582.320, env_step=84000, len=400, n/ep=0, n/st=100, qostaushy/loss=3805.256, rew=24794.00]                          


Epoch #84: test_reward: 10749.500000 ± 2703.268031, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #85: 1001it [00:13, 75.24it/s, bastaushy/loss=2185.154, env_step=85000, len=242, n/ep=0, n/st=100, qostaushy/loss=3976.371, rew=15780.00]                          


Epoch #85: test_reward: 18472.900000 ± 5660.205305, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #86: 1001it [00:13, 71.57it/s, bastaushy/loss=2422.533, env_step=86000, len=135, n/ep=0, n/st=100, qostaushy/loss=3920.253, rew=4648.00]                          


Epoch #86: test_reward: 13497.400000 ± 7193.553742, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #87: 1001it [00:13, 73.42it/s, bastaushy/loss=2682.225, env_step=87000, len=152, n/ep=1, n/st=100, qostaushy/loss=4438.722, rew=9912.00]                          


Epoch #87: test_reward: 7158.000000 ± 4508.053682, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #88: 1001it [00:13, 74.99it/s, bastaushy/loss=2237.104, env_step=88000, len=400, n/ep=0, n/st=100, qostaushy/loss=3980.785, rew=24408.00]                          


Epoch #88: test_reward: 8038.300000 ± 4016.087625, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #89: 1001it [00:13, 75.12it/s, bastaushy/loss=2915.517, env_step=89000, len=400, n/ep=0, n/st=100, qostaushy/loss=3564.886, rew=24408.00]                          


Epoch #89: test_reward: 9757.100000 ± 3851.735205, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #90: 1001it [00:13, 74.96it/s, bastaushy/loss=2540.450, env_step=90000, len=103, n/ep=0, n/st=100, qostaushy/loss=3105.911, rew=4652.00]                          


Epoch #90: test_reward: 7980.400000 ± 4946.048427, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #91: 1001it [00:13, 74.87it/s, bastaushy/loss=2984.481, env_step=91000, len=138, n/ep=2, n/st=100, qostaushy/loss=3274.651, rew=8212.00]                          


Epoch #91: test_reward: 9104.000000 ± 4583.815616, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #92: 1001it [00:13, 74.09it/s, bastaushy/loss=2664.356, env_step=92000, len=122, n/ep=1, n/st=100, qostaushy/loss=3743.359, rew=4946.00]                          


Epoch #92: test_reward: 15161.500000 ± 2857.960540, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #93: 1001it [00:13, 74.63it/s, bastaushy/loss=3035.575, env_step=93000, len=148, n/ep=0, n/st=100, qostaushy/loss=3248.003, rew=6562.00]                          


Epoch #93: test_reward: 17294.200000 ± 7523.604213, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #94: 1001it [00:13, 74.53it/s, bastaushy/loss=2837.739, env_step=94000, len=178, n/ep=0, n/st=100, qostaushy/loss=3247.127, rew=7611.00]                          


Epoch #94: test_reward: 12869.200000 ± 5125.895176, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #95: 1001it [00:14, 70.16it/s, bastaushy/loss=2687.083, env_step=95000, len=400, n/ep=0, n/st=100, qostaushy/loss=3720.616, rew=21426.00]                          


Epoch #95: test_reward: 15717.000000 ± 6917.130388, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #96: 1001it [00:13, 74.51it/s, bastaushy/loss=2599.334, env_step=96000, len=273, n/ep=0, n/st=100, qostaushy/loss=3517.812, rew=14336.00]                          


Epoch #96: test_reward: 12647.300000 ± 4524.538453, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #97: 1001it [00:13, 71.76it/s, bastaushy/loss=2866.826, env_step=97000, len=115, n/ep=1, n/st=100, qostaushy/loss=3600.567, rew=5494.00]                          


Epoch #97: test_reward: 15955.500000 ± 5917.978223, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #98: 1001it [00:13, 74.14it/s, bastaushy/loss=3055.185, env_step=98000, len=142, n/ep=0, n/st=100, qostaushy/loss=3253.939, rew=9411.00]                          


Epoch #98: test_reward: 11798.900000 ± 3944.010204, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #99: 1001it [00:13, 74.39it/s, bastaushy/loss=2642.827, env_step=99000, len=201, n/ep=0, n/st=100, qostaushy/loss=3470.781, rew=9676.00]                          


Epoch #99: test_reward: 8838.900000 ± 5139.054202, best_reward: 20748.800000 ± 2072.252195 in #52


Epoch #100: 1001it [00:13, 73.71it/s, bastaushy/loss=2455.270, env_step=100000, len=143, n/ep=0, n/st=100, qostaushy/loss=3359.213, rew=3124.00]                          


Epoch #100: test_reward: 16604.800000 ± 6628.959749, best_reward: 20748.800000 ± 2072.252195 in #52

==========Result==========
{'duration': '1446.99s', 'train_time/model': '1279.12s', 'test_step': 259246, 'test_episode': 1010, 'test_time': '140.15s', 'test_speed': '1849.78 step/s', 'best_reward': 20748.8, 'best_result': '20748.80 ± 2072.25', 'train_step': 100000, 'train_episode': 474, 'train_time/collector': '27.72s', 'train_speed': '76.52 step/s'}

(the trained policy can be accessed via policy.policies[agents[1]])


### 🐙 Evaluate best Qostaushy agent

In [ ]:
PLAYS = {"bastaushy": 0, "qostaushy": 0}
# Step 1: Load the PettingZoo environment
env = env()#render_mode="human")

# Step 2: Wrap the environment for Tianshou interfacing
env = PettingZooEnv(env)

# # Step 3: Define policies for each agent
policies = MultiAgentPolicyManager([RandomPolicy(), policy.policies[agents[1]]], env)
# # Step 4: Convert the env to vector format
env = DummyVectorEnv([lambda: env])

# # Step 5: Construct the Collector, which interfaces the policies with the vectorised environment
collector = Collector(policies, env)

# # Step 6: Execute the environment with the agents playing for 1 episode, and render a frame every ? seconds
result = collector.collect(n_episode=1000)
print(result)
print(PLAYS)

{'n/ep': 1000, 'n/st': 157498, 'rews': array([[12720.,  9676.],
       [ 2384.,   504.],
       [ 6010.,  3100.],
       ...,
       [ 2312.,   668.],
       [ 9296.,  3620.],
       [10338.,  6500.]]), 'lens': array([207,  43, 127, 281, 179,  69,  57, 143,  79,  55, 214, 147, 233,
       356, 101,  63,  55, 400, 386,  75,  53, 400, 314,  41, 215, 163,
       141, 121, 165, 227, 183, 161,  57,  63,  99, 143, 313,  81, 118,
       121, 161, 239, 161, 348,  89, 155, 125, 125,  57,  61, 159,  45,
        45, 169, 161, 169,  83, 239, 302, 326, 111, 183, 400,  93, 163,
       119, 167, 292, 191,  89,  55,  97, 141, 149, 372, 163,  79, 143,
       125, 137, 115, 336, 222, 125, 183, 111, 171, 117,  69, 331, 165,
       269,  95,  47, 113, 147, 197, 119, 183,  91,  83,  81, 103, 143,
       151,  95,  37, 269, 308, 187, 119, 107, 234,  87, 197,  65, 109,
        61,  59, 147, 263, 141,  57, 306, 259, 272, 185, 117,  93, 392,
        73,  55, 107,  91, 352,  73, 113, 101, 155,  75,  67,  49,  8

### 🐳 Experiments

1.

 net = Net(
            state_shape=observation_space.shape or observation_space.n,
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[128, 256, 256, 128],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

train_collector.collect(n_step=64 * 100)  # batch size * training_num

res: {'bastaushy': 781, 'qostaushy': 194}

res: {'bastaushy': 790, 'qostaushy': 191}

2.

 net = Net(
            state_shape=observation_space.shape or observation_space.n,
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[128, 256, 256, 128],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

train_collector.collect(n_step=256 * 100)  # batch size * training_num

res: 6/3

3.

net = Net(
            state_shape=observation_space.shape or observation_space.n,
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[128, 256, 256, 128],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

train_collector.collect(n_step=128 * 100)  # batch size * training_num     

{'bastaushy': 537, 'qostaushy': 411}

4.

net = Net(
            state_shape=observation_space.shape or observation_space.n,
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[256, 512, 512, 256],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

train_collector.collect(n_step=128 * 100)  # batch size * training_num   

res: {'bastaushy': 493, 'qostaushy': 483}

res: {'bastaushy': 512, 'qostaushy': 464}

5.

net = Net(
            state_shape=observation_space.shape or observation_space.n,
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[512, 1024, 1024, 512],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

res: {'bastaushy': 38, 'qostaushy': 59}

res: {'bastaushy': 372, 'qostaushy': 562}

### 🦎 Play with different policies

In [21]:
env = PettingZooEnv(env())
net1 = Net(
            state_shape=(22,),
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[128, 256, 256, 128],
            #hidden_sizes=[1024, 2048, 2048, 1024],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

agent1 = DQNPolicy(
            model=net1,
            optim = torch.optim.Adam(net1.parameters(), lr=1e-4),
            discount_factor=0.9,
            estimation_step=3,
            target_update_freq=320,
        ).to("cuda" if torch.cuda.is_available() else "cpu")
agent1_learned = deepcopy(agent1)
agent1_learned.load_state_dict(torch.load(agent1_path))

net2 = Net(
            state_shape=(22,),
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[256, 512, 512, 256],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

agent2 = DQNPolicy(
            model=net2,
            optim = torch.optim.Adam(net1.parameters(), lr=1e-4),
            discount_factor=0.9,
            estimation_step=3,
            target_update_freq=320,
        ).to("cuda" if torch.cuda.is_available() else "cpu")
agent2_learned = deepcopy(agent2)
agent2_learned.load_state_dict(torch.load(agent2_path))

net3 = Net(
            state_shape=(22,),
            action_shape=env.action_space.shape or env.action_space.n,
            hidden_sizes=[512, 1024, 1024, 512],
            device="cuda" if torch.cuda.is_available() else "cpu",
        ).to("cuda" if torch.cuda.is_available() else "cpu")

agent3 = DQNPolicy(
            model=net3,
            optim = torch.optim.Adam(net3.parameters(), lr=1e-4),
            discount_factor=0.9,
            estimation_step=3,
            target_update_freq=320,
        ).to("cuda" if torch.cuda.is_available() else "cpu")

agent3_learned = deepcopy(agent3)
agent3_learned.load_state_dict(torch.load(agent3_path))

<All keys matched successfully>

In [58]:
PLAYS = {"bastaushy": 0, "qostaushy": 0}
# Step 1: Load the PettingZoo environment
env = env()#render_mode="human")

# Step 2: Wrap the environment for Tianshou interfacing
env = PettingZooEnv(env)

# # Step 3: Define policies for each agent
policies = MultiAgentPolicyManager([agent1_learned, agent2_learned], env)
# # Step 4: Convert the env to vector format
env = DummyVectorEnv([lambda: env])

# # Step 5: Construct the Collector, which interfaces the policies with the vectorised environment
collector = Collector(policies, env)

# # Step 6: Execute the environment with the agents playing for 1 episode, and render a frame every ? seconds
result = collector.collect(n_episode=100)
print(PLAYS)

{'bastaushy': 100, 'qostaushy': 0}


🐯 Experiment results


*   agent3 vs agent1: 0-0
*   agent1 vs agent3: 0-100
*   agent3 vs agent2: 0-100
*   agent2 vs agent3: 100-0
*   agent2 vs agent1: 0-100
*   agent1 vs agent2: 100-0

